src/org/etools/j1939_84/controllers/part1/BusService.java

```diff
@@ -0,0 +1,188 @@
+     *         the number of seconds to read the bus
+     * @return the List of GenericPackets that were received
+     */
+    public List<GenericPacket> readBus(int seconds, Predicate<GenericPacket> filter) {
+        listener.onResult("Reading bus for " + seconds + " seconds");
+        long stopTime = System.currentTimeMillis() + seconds * 1000L;
+        new Thread(() -> {
+            long secondsToGo = seconds;
+            while (secondsToGo > 0) {
+                secondsToGo = (stopTime - System.currentTimeMillis()) / 1000;
+                listener.onProgress(String.format("Reading bus for %1$d seconds", secondsToGo));
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException ignored) {
+                }
+            }
+        }).start();
+
+        return j1939.read(GenericPacket.class, seconds, TimeUnit.SECONDS)
+                .flatMap(e -> e.left.stream())
+                .filter(filter)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Sends a Request to the Global Address for the given PGN
+     */
+    public List<GenericPacket> globalRequest(int pgn) {
+        listener.onResult(J1939_84.NL);
+        String title = "Global Request for " + pgn;
+        Packet requestPacket = j1939.createRequestPacket(pgn, J1939.GLOBAL_ADDR);
+        RequestResult<GenericPacket> globalResult = j1939.requestGlobal(title,
+                                                                        listener,
+                                                                        true,
+                                                                        GenericPacket.class,
+                                                                        pgn,
+                                                                        requestPacket);
+        return globalResult
+                .getEither()
+                .stream()
+                .flatMap(e -> e.left.stream())
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Sends a Destination Specific Request for the given PGN to the given moduleAddress
+     *
+     * @param pgn
+     *         the PGN of interest
+     * @param moduleAddress
+     *         the module address of interest
+     * @return List of Packets received
+     */
+    public List<GenericPacket> dsRequest(int pgn, int moduleAddress) {
+        listener.onResult(J1939_84.NL);
+        String title = "DS Request for " + pgn + " to " + Lookup.getAddressName(moduleAddress);
+        Packet requestPacket = j1939.createRequestPacket(pgn, moduleAddress);
+        return j1939.requestDS(title,
+                               listener,
+                               true,
+                               GenericPacket.class,
+                               pgn,
+                               requestPacket)
+                .getPacket()
+                .stream()
+                .filter(p -> p.left.isPresent())
```

Comment:
Does it make sense to ignore NACKs?

Reply:
This method is specifically designed its current use.  If we need to to also get back the NACKs in the same stream, because we find it useful, we can update the logic here without negatively affecting the current use case.